Automattic/woocommerce-payments
# Show last 4 digit credit card in subscriptions order's note when updating failed payment

Fixes #871

#### Changes proposed in this Pull Request

* Add filter hooks to ones defined in this 3908-gh-woocommerce/woocommerce-subscriptions .
* Get _old_ payment's last 4 digit from 2nd to last payment token because payment tokens are added in chronological order, so the old (failing) payment is the 2nd to last in that order's payment tokens array.
* Get _new_ payment's last 4 digit from last payment token or from request, depending whether user choose from a saved payment or a new payment.

#### Testing instructions



Follow these steps, [copied from PR#832](https://github.com/Automattic/woocommerce-payments/pull/832):
1. Place an order for a free-trial subscription product or update an existing subscription with an attachable but declined card, such as `4000000000000341`
2. Assert that the order or update is successful
4. Go to the subscriptions admin page and click on the subscription
5. Run the `Process renewal` action
6. Assert that the renewal fails
7. Go to My Account > Subscriptions and click on the subscription
8. Scroll down to the order list and click on Pay for the failed order
9. Insert another basic card details (or use a saved card) and update the failed order
10. Assert that the payment for the renewal order is successful
11. Go to the subscriptions admin page, click on the subscription and run the `Process renewal` action
13. Assert that the renewal used the card used in the failed renewal and that the transaction is listed in the WCPay transactions page 

Notice the order note of the subscription order will show:
```Payment method changed from “Credit card” to “Credit card” by the subscriber from their account page.```

With this fix, it should show:
<img width="305" alt="Screen Shot 2020-12-07 at 08 48 58" src="https://user-images.githubusercontent.com/73803630/101300754-2fd2d580-3869-11eb-9226-f41ff6e7b456.png">


-------------------
